### PR TITLE
add VEP process management

### PR DESCRIPTION
**Pull request by sheridancbio**
- SystemProcessManager class now provides process management services for VEP
        - uses UNIXProcess (if available) for determining a Process pid
        - uses UNIXProcess (if available) for reaping zombies (via waitForProcessExit())
        - determines availability of ps and kill commands on local system
        - manages processes launched if needed tools are available
        - starts ReaperDaemonThread when first VEP process is launched
        - reaper will periodically check for orphaned child processes of a destroyed VEP process and kill them
        - reaper will periodically check for zombie processes and reap them
    - unix 'ps' command output parsing encapsulated in ProcessSurveyor class
    - remove thread interruption in VepRunner - now StreamTransferrer threads will exit naturally when orphaned children are killed
    - shut down reaper as application shuts down

### CODE REVIEW

File: src/main/java/org/genomenexus/vep_wrapper/ProcessSurveyor.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        return new ProcessSurveyItem(processId, parentProcessId, stateCode, fields[firstFieldIndex + 3]);
+    }
+
+    public static List<ProcessSurveyItem> getProcessSurvey() {
+        ArrayList<ProcessSurveyItem> processSurvey = new ArrayList<ProcessSurveyItem>();
+        try {
+            Process psProcess = psProcessBuilder.start();
+            BufferedReader in = new BufferedReader(new InputStreamReader(psProcess.getInputStream()));
+            while (true) {
+                String line = in.readLine();
```
Comment (ao508) [inline src/main/java/org/genomenexus/vep_wrapper/ProcessSurveyor.java:77]: isn't it equivalent to use:

```
while ((line = in.readLine()) != null)
...
```

?

I am hesitant to use `true` in the while statement

Comment (ao508) [reply src/main/java/org/genomenexus/vep_wrapper/ProcessSurveyor.java:77]: i suppose the item being parsed would also need to be in the while statement check as well

File: src/main/java/org/genomenexus/vep_wrapper/SystemProcessManager.java [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+    public void run() {
+        if (psCommandAvailableOnSystem == null) {
+            determineIfPsCommandIsAvailable();
+        }
+        if (killCommandAvailableOnSystem == null) {
+            determineIfKillCommandIsAvailable();
+        }
+        // ReaperDaemonThread
+        while (true) {
```
Comment (ao508) [inline src/main/java/org/genomenexus/vep_wrapper/SystemProcessManager.java:126]: hesitant to use `true` here as well but at the same time there's a lot going on in this block so maybe just keeping it `true` is best to keep it simple?

File: src/main/java/org/genomenexus/vep_wrapper/ProcessSurveyor.java (hunk 1 above)
Comment (sheridancbio) [reply src/main/java/org/genomenexus/vep_wrapper/ProcessSurveyor.java]: fixed as you suggested

File: src/main/java/org/genomenexus/vep_wrapper/SystemProcessManager.java (hunk 2 above)
Comment (sheridancbio) [reply src/main/java/org/genomenexus/vep_wrapper/SystemProcessManager.java]: I have fixed this too. The reaper thread, once started, will continue to run as long as the application is running. So now the condition of the while statement checks whether a shutdown of the reaper was requested and exits if it sees such a request.

I have added a servlet configuration bean which will send the request for reaper shutdown as the servlet is exiting ... and will give a few seconds for the reaper to complete its work before exiting the application.

**Comment by sheridancbio** [comment]
Squashing and updating commit message

**Pull request merged**